### ISSUE TITLE
Protocol error (Runtime.callFunctionOn): Cannot find context with specified id undefined

### ISSUE THREAD

**Issue by bezoerb** [pocketjoso/penthouse#263]
Using the following html causes an error in `lib/browser-sandbox/pruneNonCriticalSelectors.js`

Debug output: 
```
...
penthouse:core page load DONE +30ms
penthouse:core pruneNonCriticalSelector threw an error: Error: Protocol error (Runtime.callFunctionOn): Cannot find context with specified id undefined +351ms
...
penthouse remove browser page for generateCriticalCss after ERROR, now: 0 +732ms
```

HTML:
```
<!doctype html>
<html>

<head>
    <title>https://example.com/</title>
    <link rel="canonical" href="https://example.com/" />
    <meta name="robots" content="noindex">
    <meta http-equiv="content-type" content="text/html; charset=utf-8" />
    <!--
    The next meta makes critical (or puppeteer) throw:
    Protocol error (Runtime.callFunctionOn): Cannot find context with specified id undefined
    -->
    <meta http-equiv="refresh" content="0; url=https://example.com/" />
</head>

</html>
```

It looks like ```<meta http-equiv="refresh" content="0; url=https://example.com/" />``` is the *bad* line.

See https://github.com/addyosmani/critical/issues/314

**Comment by XhmikosR** [comment]
Thanks for tracking down the root cause @bezoerb.

/CC @pocketjoso in case you are not watching the repo :)

**Comment by XhmikosR** [comment]
So it seems that catching the Promise error in https://github.com/pocketjoso/penthouse/blob/master/src/core.js#L345 fixes the error, but I'm not sure if the catch should be a generic one or not.

Also, I notice with the catch there, then critical inlines the critical CSS (from the other pages) even though that page has no stylesheets at all. So, it's not the best solution :/

**Comment by pocketjoso** [comment]
Thanks @XhmikosR - I am watching, I was just on vacation.
Not sure what we want to do here. Even if the error was caught, penthouse should still output an error imo. We could just return a more specific error, but would that help much?

**Comment by XhmikosR** [comment]
Unfortunately, no. Because the pages with the meta refresh tag are generated by my static site generator. I could filter them out,  by keeping track of a huge list of HTML files to ignore, or we could make pockethouse not throw. Are there any potential cases `pruneNonCriticalSelectors` can throw?

**Comment by hummal** [comment]
I had this problem long time ago in #202. You can read there if you want. 

After I had tracked down and fixed the problem I made a pr. Thats all of the history. The rest is in the ticket.

**Comment by pocketjoso** [comment]
@hummal are you sure your PR handled meta refresh tags?

I still think blocking navigation would be the best option (and generate critical css for the first page), but this does not seem easy. In-browser JS cannot override the meta refresh navigation, and currently Puppeteer throws if calling `abort` on a navigation request. I'll raise an issue there and see further if this is possible in another way.

The other alternative is to catch the error.. but @XhmikosR I don't see any better option than throwing a specific well documented error, that you can catch and ignore. Penthouse could not just return `""` - you would still need to handle this case (make sure you don't end up serving that as critical css).

> Are there any other cases pruneNonCriticalSelectors can throw?

Probably, but how do you think this is relevant? The error raised here comes from trying to evaluate JS on a page that no longer exists, as it has been navigated away from. It doesn't relate specifically to the `pruneNonCriticalSelectors` code.

**Comment by XhmikosR** [comment]
I guess this could be handled in critical.

**Comment by hummal** [comment]
@pocketjoso no it doesn't but the problem here is maybe another than all of you expect. Also it is clearly NOT an navigation problem. I tracked that down with nginx server log analysis and it happens only when the status code is 499 "Client closes request". It is rare but it happens. Thats why I commented and fixed that problem in the pr with changing the moment when to stop page request after loading timeout. 

But maybe Iam wrong. At the moment I never had such an issue with crittr

**Comment by pocketjoso** [comment]
@hummal The error reported here happens in Puppeteer when you try to call `page.evaluate` when there is no loaded page present. It can easily be reproduced with the HTML provided in the first post. It sounds like you were hitting on other, possibly related errors. If the server closes the request before the page has been able to fully load I would expect this to lead to a timeout in Penthouse, not the error reported here - but its not an error path intentionally handled so I'm not quite sure.

@XhmikosR I kind of feel that would be better. I realised Penthouse would otherwise also need to distinguish between cases where a meta refresh is used to refresh the _same_ page (get latest queue position or similar, nowadays much better done via ~AJAX), and where it's essentially a redirect to _another_ page (better done on the server). In the first case we would want to disable the redirect and generate critical css for the first page. In the other case, if anything we would want critical css for the second page (redirected to), if it's even a page we own. This kind of behaviour is not easy to handle generically in a library like Penthouse.

**Comment by XhmikosR** [comment]
@pocketjoso: in for critical to detect this case, shouldn't pockethouse emit something different like you mention above?

**Comment by pocketjoso** [comment]
Sure, something like this?
```
PAGE_UNLOADED: Critical css generation script could not be executed.

This most often happens when the page is navigated away from after load, f.e. via a meta tag refresh directive.
```

**Comment by XhmikosR** [comment]
Sounds good to me.

Note that the same happens if one uses `window.location='foo'`.

**Comment by pocketjoso** [comment]
Updated error description in previous post to mention window.location setting, changed error name to `PAGE_UNLOADED_DURING_EXECUTION`. It's quite natural that this would error if you think of it, since the page the script is execution on unloads (new url is loaded).

**Comment by pocketjoso** [comment]
Testing this under `penthouse@1.6.1-rc` (tag `rc`). Will release as a patch when I've checked it doesn't cause regressions, most likely tomorrow.

**Comment by pocketjoso** [comment]
Available as `1.6.1` now (latest) - 
https://github.com/pocketjoso/penthouse/releases/tag/v.1.6.1
(this was the only change)
